Error in offline validation when using additional annotations

**Description**

When using annotations, I am able to validate them by using the signature being stored in the registry but not when using the offline validation capability. 

For being able to recreate the tasks I did:

cosign sign --upload=false --key awskms:///arn:aws:kms:eu-west-3:xxxxxx:alias/cosign-key  -a appname=myapp   --output-signature=mysig1.sig xxx/yyy:latest

cosign verify -a appname=myapp   --key cosign.pub --signature mysig1.sig  xxx/yyy:latest

I'm getting the following error message:
WARNING: using obsolete implied signature payload data (with digested reference index.docker.io/xxx/bla@sha256:db071ebcec3e74bfb9a6e0358a233f7b4cc38585d3201239b9239d2e287d7e9a); specify it explicitly with --payload instead
Error: no matching signatures: searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: crypto/rsa: verification error}
main.go:69: error during command execution: no matching signatures: searching log query: [POST /api/v1/log/entries/retrieve][400] searchLogQueryBadRequest  &{Code:400 Message:verifying signature: crypto/rsa: verification error}

When doing the validation online without having specified --upload=false or not specifying any annotation and doing then the offline validation by using --signature=mysig1.sig the validation is successful. 

I am not sure if this is actually a bug but when reading the documentation my expectation is that cosign verify -a appname=myapp   --key cosign.pub --signature mysig1.sig  xxx/yyy:latest should work in general.

**Version**

  ______   ______        _______. __    _______ .__   __.
 /      | /  __  \      /       ||  |  /  _____||  \ |  |
|  ,----'|  |  |  |    |   (----`|  | |  |  __  |   \|  |
|  |     |  |  |  |     \   \    |  | |  | |_ | |  . `  |
|  `----.|  `--'  | .----)   |   |  | |  |__| | |  |\   |
 \______| \______/  |_______/    |__|  \______| |__| \__|
cosign: A tool for Container Signing, Verification and Storage in an OCI registry.

GitVersion:    2.4.1
GitCommit:     9a4cfe1aae777984c07ce373d97a65428bbff734
GitTreeState:  "clean"
BuildDate:     2024-10-03T17:01:50Z
GoVersion:     go1.23.2
Compiler:      gc
Platform:      darwin/arm64

--- a/cmd/cosign/cli/attach/sig.go
+++ b/cmd/cosign/cli/attach/sig.go
@@ -55,7 +55,7 @@ func SignatureCmd(ctx context.Context, regOpts options.RegistryOptions, sigRef, 
 
 	var payload []byte
 	if payloadRef == "" {
-		payload, err = cosign.ObsoletePayload(ctx, digest)
+		payload, err = cosign.ObsoletePayload(ctx, digest, nil)
 	} else {
 		payload, err = os.ReadFile(filepath.Clean(payloadRef))
 	}

--- a/pkg/cosign/obsolete.go
+++ b/pkg/cosign/obsolete.go
@@ -26,8 +26,8 @@ import (
 // ObsoletePayload returns the implied payload that some commands expect to match
 // the signature if no payload is provided by the user.
 // DO NOT ADD ANY NEW CALLERS OF THIS.
-func ObsoletePayload(ctx context.Context, digestedImage name.Digest) ([]byte, error) {
-	blob, err := (&payload.Cosign{Image: digestedImage}).MarshalJSON()
+func ObsoletePayload(ctx context.Context, digestedImage name.Digest, annotations map[string]interface{}) ([]byte, error) {
+	blob, err := (&payload.Cosign{Image: digestedImage, Annotations: annotations}).MarshalJSON()
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/cosign/obsolete_test.go
+++ b/pkg/cosign/obsolete_test.go
@@ -32,10 +32,26 @@ func TestObsoletePayload(t *testing.T) {
 	require.NoError(t, err)
 	var res []byte
 	stderr := ui.RunWithTestCtx(func(ctx context.Context, _ ui.WriteFunc) {
-		r, err := ObsoletePayload(ctx, digestedImg)
+		r, err := ObsoletePayload(ctx, digestedImg, nil)
 		require.NoError(t, err)
 		res = r
 	})
 	assert.Contains(t, stderr, "obsolete implied signature payload")
 	assert.Equal(t, []byte(`{"critical":{"identity":{"docker-reference":"index.docker.io/namespace/image"},"image":{"docker-manifest-digest":"sha256:4aa3054270f7a70b4528f2064ee90961788e1e1518703592ae4463de3b889dec"},"type":"cosign container image signature"},"optional":null}`), res)
+}
+
+func TestObsoletePayloadWithAnnotations(t *testing.T) {
+	// The implied payload must include annotations passed by the caller so it matches
+	// the payload generated when signing with annotations (cmd/cosign/cli/generate).
+	digestedImg, err := name.NewDigest("docker.io/namespace/image@sha256:4aa3054270f7a70b4528f2064ee90961788e1e1518703592ae4463de3b889dec")
+	require.NoError(t, err)
+	annotations := map[string]interface{}{"appname": "myapp"}
+	var res []byte
+	stderr := ui.RunWithTestCtx(func(ctx context.Context, _ ui.WriteFunc) {
+		r, err := ObsoletePayload(ctx, digestedImg, annotations)
+		require.NoError(t, err)
+		res = r
+	})
+	assert.Contains(t, stderr, "obsolete implied signature payload")
+	assert.Equal(t, []byte(`{"critical":{"identity":{"docker-reference":"index.docker.io/namespace/image"},"image":{"docker-manifest-digest":"sha256:4aa3054270f7a70b4528f2064ee90961788e1e1518703592ae4463de3b889dec"},"type":"cosign container image signature"},"optional":{"appname":"myapp"}}`), res)
 }

--- a/pkg/cosign/verify.go
+++ b/pkg/cosign/verify.go
@@ -1040,7 +1040,7 @@ func loadSignatureFromFile(ctx context.Context, sigRef string, signedImgRef name
 		if err != nil {
 			return nil, err
 		}
-		payload, err = ObsoletePayload(ctx, digest)
+		payload, err = ObsoletePayload(ctx, digest, co.Annotations)
 		if err != nil {
 			return nil, err
 		}

